lib/astutils.cpp

```diff
@@ -145,6 +145,12 @@ bool isSameExpression(bool cpp, bool macro, const Token *tok1, const Token *tok2
         if (tok2->str() == "." && tok2->astOperand1() && tok2->astOperand1()->str() == "this")
             tok2 = tok2->astOperand2();
     }
+    if(Token::simpleMatch(tok1, "! !")) {
```

Comment:
I would suggest that the AST is used. This code does not handle `! ! (x+y)` etc.